File: docs/analisis_data_v2.4/e2e_best_practice/src/frontend/src/services/timeConverter.js
/**
 * Utilitas Konversi Waktu Otomatis dengan Indikator Akurasi
 * Berdasarkan: docs/analisis data v2/global/aturan_konversi_waktu_V2.1.md
 */

export const TIME_UNITS = {
  MONTH: 'month',
  DAY: 'day',
  HOUR: 'hour',
  MINUTE: 'minute',
  SECOND: 'second'
};

const THRESHOLDS = {
  [TIME_UNITS.MONTH]: { sub: TIME_UNITS.DAY, size: 30 },
  [TIME_UNITS.DAY]: { sub: TIME_UNITS.HOUR, size: 24 },
  [TIME_UNITS.HOUR]: { sub: TIME_UNITS.MINUTE, size: 60 },
  [TIME_UNITS.MINUTE]: { sub: TIME_UNITS.SECOND, size: 60 }
};

/**
 * Mengonversi nilai antar skala waktu dan menghitung akurasi data.
 * 
 * @param {number|number[]} nilai - Nilai tunggal atau array nilai untuk diagregasi (AVG)
 * @param {string} satuanAsal - Satuan waktu asal (second, minute, hour, day, month)
 * @param {string} satuanTujuan - Satuan waktu tujuan (second, minute, hour, day, month)
 * @returns {Object} Hasil konversi dan informasi akurasi
 */
export function convertTimeScale(nilai, satuanAsal, satuanTujuan) {
  // 1. Validasi Input
  const validUnits = Object.values(TIME_UNITS);
  if (!validUnits.includes(satuanAsal) || !validUnits.includes(satuanTujuan)) {
    throw new Error(`Satuan tidak valid: ${satuanAsal} -> ${satuanTujuan}`);
  }

  // Normalisasi input ke array
  const dataArray = Array.isArray(nilai) ? nilai : [nilai];
  if (dataArray.length === 0) {
    return { value: 0, accuracy: '0% akurat', isWarning: true, message: 'Data kosong' };
  }

  // Hitung rata-rata awal
  const avgValue = dataArray.reduce((a, b) => a + b, 0) / dataArray.length;

  // Jika satuan sama
  if (satuanAsal === satuanTujuan) {
    return {
      value: avgValue,
      accuracy: '100% akurat',
      isWarning: false,
      message: 'Satuan sama, tidak ada konversi'
    };
  }

  // 2. Logika Konversi Otomatis
  const targetThreshold = THRESHOLDS[satuanTujuan];

  // Cek apakah konversi 1 tingkat ke atas (misal: day -> month)
  if (targetThreshold && targetThreshold.sub === satuanAsal) {
    const T = targetThreshold.size;
    const count = dataArray.length;
    const accuracyPct = Math.min(100, Math.round((count / T) * 100));
    
    return {
      value: avgValue,
      accuracy: `${accuracyPct}% akurat`,
      isWarning: accuracyPct < 100,
      message: accuracyPct < 100 
        ? `telah dikonversi hanya ${accuracyPct}% akurat` 
        : '100% akurat'
    };
  }

  // 3. Penanganan konversi multi-level (Opsional/Fallback)
  // Untuk saat ini hanya mendukung 1 tingkat sesuai permintaan aturan
  return {
    value: avgValue,
    accuracy: 'Konversi tidak didukung',
    isWarning: true,
    message: `Sistem saat ini hanya mendukung konversi 1 tingkat (misal: ${satuanAsal} ke ${targetThreshold?.sub || 'satuan di atasnya'})`
  };
}

File: project_backup_20260305/backend/app/core/time_converter.py
from enum import Enum
from typing import List, Union, Dict, Any

class TimeUnits(str, Enum):
    MONTH = "month"
    DAY = "day"
    HOUR = "hour"
    MINUTE = "minute"
    SECOND = "second"

THRESHOLDS = {
    TimeUnits.MONTH: {"sub": TimeUnits.DAY, "size": 30},
    TimeUnits.DAY: {"sub": TimeUnits.HOUR, "size": 24},
    TimeUnits.HOUR: {"sub": TimeUnits.MINUTE, "size": 60},
    TimeUnits.MINUTE: {"sub": TimeUnits.SECOND, "size": 60}
}

def convert_time_scale(
    values: Union[float, List[float]], 
    unit_from: str, 
    unit_to: str
) -> Dict[str, Any]:
    """
    Logika Konversi Waktu Otomatis (Backend Version)
    Berdasarkan: docs/analisis data v2/global/aturan_konversi_waktu_V2.1.md
    """
    # 1. Validasi Input
    try:
        u_from = TimeUnits(unit_from)
        u_to = TimeUnits(unit_to)
    except ValueError:
        raise ValueError(f"Satuan tidak valid: {unit_from} -> {unit_to}")

    # Normalisasi input ke list
    data_list = values if isinstance(values, list) else [values]
    if not data_list:
        return {
            "value": 0.0,
            "accuracy": "0% akurat",
            "is_warning": True,
            "message": "Data kosong"
        }

    # Hitung rata-rata awal
    avg_value = sum(data_list) / len(data_list)

    # Jika satuan sama
    if u_from == u_to:
        return {
            "value": avg_value,
            "accuracy": "100% akurat",
            "is_warning": False,
            "message": "Satuan sama, tidak ada konversi"
        }

    # 2. Logika Konversi Otomatis
    target_threshold = THRESHOLDS.get(u_to)

    # Cek apakah konversi 1 tingkat ke atas (misal: day -> month)
    if target_threshold and target_threshold["sub"] == u_from:
        T = target_threshold["size"]
        count = len(data_list)
        accuracy_pct = min(100, round((count / T) * 100))
        
        return {
            "value": avg_value,
            "accuracy": f"{accuracy_pct}% akurat",
            "is_warning": accuracy_pct < 100,
            "message": f"telah dikonversi hanya {accuracy_pct}% akurat" if accuracy_pct < 100 else "100% akurat"
        }

    # 3. Penanganan konversi multi-level (Fallback)
    return {
        "value": avg_value,
        "accuracy": "Konversi tidak didukung",
        "is_warning": True,
        "message": f"Sistem saat ini hanya mendukung konversi 1 tingkat (misal: {unit_from} ke {target_threshold['sub'] if target_threshold else 'satuan di atasnya'})"
    }
